### PR TITLE
Splitting Cats and Dogs

### PR DESCRIPTION
**Pull request by nixpulvis**
This post is partially a response to some discussion at work, and partially a dump of some of my thoughts on the topic of modularization and code sharing.

I'd love feedback on what makes sense, what resonates, and what is missing. If you agree, disagree or simply find spelling/grammar mistakes I'd **really** love for you to let me know.

**The rendered version of this is drafted here: https://nixpulvis.com/ramblings/2019-01-05-splitting-cats-and-dogs, I recommend reading it there instead of in this PR.**

### CODE REVIEW

**Comment by nixpulvis** [comment]
I want to add a few pictures like this for both "Grouping" and "Sharing" too...

image from ios

I just need to learn how to use some kind of vector animation software...

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+might be looking at a class with the need for countless functions. Instead of
+defining them all one after another, it makes sense to _group_ the functions
+and move them somewhere so it's clear how they are related. Or, you might be
+looking to _share_ some logic across a few classes. This is key to writing
+good software, since maintaining two versions of the same thing is cumbersome.
+Remember, stay DRY (Don't Repeat Yourself).
+
+### Grouping
+
+Grouping shouldn't require too much thought, aside from a **name** for the way
```
Comment (rbritom) [inline _ramblings/2019-01-05-splitting-cats-and-dogs.md]: the argument here is an over simplification, this here only implied cohesion, but you also need to think about decoupling and dependency management, try to see if this concepts and be alluded briefly somehow.

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+Often when you're programming you find the need to split things up a bit. This
+is such a fundamental part of programming that almost every language gives you
+many tools to aid with this. Maybe it's _modules_, _classes_, _impls_, or
+even _packages_. The list goes on... In fact, one of the best forms of
+splitting can be a simple _function_. Even _comments_ can play a role in this
+if you'd like.
+
+There's a number of reasons you'd want to split up some part of a program. You
+might be looking at a class with the need for countless functions. Instead of
```
Comment (rbritom) [inline _ramblings/2019-01-05-splitting-cats-and-dogs.md:15]: To me splitting code is also about run time complexity, not just readability, an object that is huge and bloated when I am in the debugger its very hard to keep track of, this is not related to the examples you present here, its just pointing out that this might be an over simplification.

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+  include Communicative
+  include Consumptive
+end
+```
+
+People often view a `module` as a way to share functionality, but here we're
+just using it to group similar functions. The main thing we've done is make it
+clear which functions are related, and therefor **a good idea to read
+together**.  If you want to share some part of your program's design, then
```
Comment (rbritom) [inline _ramblings/2019-01-05-splitting-cats-and-dogs.md:139]: exactly

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+    end
+
+    @stomach = @stomach - victuals
+  end
+end
+```
+
+Notice how we've made use of a Ruby instance variable here. It's a really good
+idea to keep the use of these as close to each other as possible. It's very
+hard to reason about a module that's touching another module's state. Keep
```
Comment (rbritom) [inline _ramblings/2019-01-05-splitting-cats-and-dogs.md:201]: I think that is a sign that it should be a class instead

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    assert_equal @dog.bark, 'Woof'
+  end
+
+  # Tests ommited...
+end
+```
+
+Finally, we can test the `Consumptive` module. We want to test this
+functionality without relying on our `Dog` or `Cat` since they aren't the
+concern of this module. A dummy, class can be created just for our test, making
```
Comment (rbritom) [inline _ramblings/2019-01-05-splitting-cats-and-dogs.md]: I think you dont need to the comma after dummy

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md (hunk 2 above)
Comment (nixpulvis) [reply _ramblings/2019-01-05-splitting-cats-and-dogs.md:15]: @rbritom I'm not disagreeing with you, but I wouldn't call what you're talking about "runtime" complexity, maybe it's cognitive complexity, or I might be tempted to just yell at the damn asswipe that designed it that way.

Runtime complexity should be talking about the performance characteristics.

File: _ramblings/2019-01-05-splitting-cats-and-dogs.md (hunk 1 above)
Comment (nixpulvis) [reply _ramblings/2019-01-05-splitting-cats-and-dogs.md]: But do we really need to bring all these terms into play, I'm leaving a lot implict because frankly it's not critical. How does dependency management play into this here, it's very subtle. The only management required is done from within Ruby's internals, and most languages do the same kinds of things. The fact that namespace resolution works is neat, and a fun topic in it of itself, but frankly outside the scope of this.

I've tweaked the language of this sentence to mention cohesion off-hand, as I do feel this is a good descriptor.

**Comment by JJPro** [comment]
Misspelling in last para of **Grouping**: 
> The main thing we’ve done is make it clear which functions are related, 
> and _**therefor**_ a good idea to read together. 

**therefore**
